app/models/concerns/omniauthable.rb

```diff
@@ -0,0 +1,83 @@
+# frozen_string_literal: true
+
+module Omniauthable
+  extend ActiveSupport::Concern
+
+  TEMP_EMAIL_PREFIX = 'change@me'
+  TEMP_EMAIL_REGEX = /\Achange@me/
+
+  included do
+    def omniauth_providers
+      providers = []
+      providers << :cas if ENV['CAS_ENABLED']
+      providers
+    end
+
+    def email_verified?
+      email && email !~ TEMP_EMAIL_REGEX
+    end
+  end
+
+  class_methods do
+    def find_for_oauth(auth, signed_in_resource = nil)
+      # EOLE-SSO Patch
+      auth.uid = (auth.uid[0][:uid] || auth.uid[0][:user]) if auth.uid.is_a? Hashie::Array
+
+      # Get the identity and user if they exist
+      identity = Identity.find_for_oauth(auth)
+
+      # If a signed_in_resource is provided it always overrides the existing user
+      # to prevent the identity being locked with accidentally created accounts.
+      # Note that this may leave zombie accounts (with no associated identity) which
+      # can be cleaned up at a later date.
+      user = signed_in_resource ? signed_in_resource : identity.user
+
+      # Create the user if needed
+      user = find_or_create_user_for_oauth(auth) if user.nil?
+
+      # Associate the identity with the user if needed
+      if identity.user != user
+        identity.user = user
+        identity.save!
+      end
+
+      user
+    end
+
+    def find_or_create_user_for_oauth(auth)
+      # Get the existing user by email if the provider gives us a verified email.
+      # If no verified email was provided we assign a temporary email and ask the
+      # user to verify it on the next step via UsersController.finish_signup
+      email_is_verified = auth.info.email && (auth.info.verified || auth.info.verified_email)
+      email = auth.info.email if email_is_verified
+      uid = auth.uid
+      user = Account.find_by(username: uid).try(:user) if uid
```

Comment:
@immae ok i see. I'm not familiar with such mechanism, but it doesn't seem to be too painful to implement. So now the question is :

The most common use scenario for CAS or SAML authentication is for organizations that already have an LDAP directory somewhere and just want to "plug" Mastodon to their existing SSO system. To do so, they will most of the time disable Mastodon database registration. Furthermore, they will use CAS or SAML, not both. So, is there really a need for reconciliation feature at this point ? Couldn't we just open an new "enhancement" issue for now and take care of that when we'll actually implement some oauth strategy that requires this feature (Facebook, Twitter, Google, Github...) ?